### ISSUE TITLE
[java] UnnecessaryLocalBeforeReturn/VariableCanBeInlined: deprecate property statementOrderMatters

### ISSUE THREAD

**Issue by adangel** [pmd/pmd#3124]
**Affects PMD Version:** 6.x

**Rule:**
[UnnecessaryLocalBeforeReturn](https://pmd.github.io/latest/pmd_rules_java_codestyle.html#unnecessarylocalbeforereturn)

**Description:**

The property `statementOrderMatters` is not so helpful. If it is set to false, the rule is relaxed to ignore any statements between the local variable assignment and the return statement under the assumption, that these statements have no side-effects. This is just too simple. We should assume, that all methods calls could have side-effects and reordering the local var assignment could lead to different results. That means, by fixing a PMD violation, a application bug could be introduced.

**Code Sample demonstrating the issue:**

```java
public class Foo {
    public int noViolation() {
        int res = calcValue(); // might throw, might have a side-effect for doSomething()
        doSomething();
        return res;
    }
    public int violation() { // can be reordered, because res is not used somewhere else
        int res = 2; // a constant
        doSomething();
        return res;
    }
    public int noViolation2() {
        int res = 2; // a constant
        doSomething(res);
        return res;
    }
    public void doSomething() {
        // omitted
    }
    public void doSomething(int c) {
        // omitted
    }
    public int calcValue() {
        // omitted
    }
}
```

**Expected outcome:**

*   In the end, the rule should be smart enough to not produce false-positive/false-negatives even without this property.

**Related issue:**

* #310
* https://github.com/pmd/pmd/pull/3113#discussion_r579216253

**Comment by adangel** [comment]
Note - the rule UnnecessaryLocalBeforeReturn has been deprecated with 7.17.0 and the replacement rule is called VariableCanBeInlined

See #5770 / #5847

**Comment by fudianchn** [comment]
Current state on main (`6dbb11d`):

- UnnecessaryLocalBeforeReturn is deprecated since 7.17.0 and its description points to VariableCanBeInlined as the replacement (#5770 / #5847).
- VariableCanBeInlined declares `statementOrderMatters` with `defaultValue(true)`, so the conservative behavior is now the default.
- Checked with the sample from this issue: on 6.31.0 with `statementOrderMatters=false`, the `noViolation()` method is reported (the unsafe suggestion this issue is about); on main, both the deprecated rule with its default and VariableCanBeInlined with its default report nothing there, while the consecutive case `int res = 2; return res;` is still reported by VariableCanBeInlined.

Can this be closed?

**Comment by UncleOwen** [comment]
No, I don't think `statementOrderMatters=false` makes any sense and should be deprecated on the new rule.
